### PR TITLE
[GEOT-5623]: dynamic stroke-dasharray improvements

### PR DESCRIPTION
**Pull request by wumpz**
Hopefully I have adressed all complaints about the older pull request (#1467). 
Now the PR is against master.
I have removed all formatting issues (that was hard).
I have as far as I see it now a more general version.

### CODE REVIEW

**Comment by wumpz** [comment]
As I see there are two commits, I hope the squash possibilities of github are enough to provide a single one.

**Comment by nmco** [comment]
Thanks for the hard work on this one, I will try to review this next weekend, we are in the middle of the code freeze anyway.

Regarding the commit squashing, it can be done by the merger but the commit will look like it's done
by the merger not by you. You will lose attribution of this work.

You can squash you commits by doing ``git rebase --interactive HEAD~2`` and select the commits you want to suqash. Then you will need to force a push using ``git push -f``. 

But this is up to you, as I say the merger can squash them when merging.

**Comment by wumpz** [comment]
I would like the attribution, but as you mentioned it was hard to get here and I have no time left to get any futher, at least at the moment. More important to me is the merge into GeoTools. Nevertheless I do not understand GeoTools one commit PR policy, as far as I know.

**Comment by aaime** [comment]
The one commit policy is strict only if something need backporting later, as it makes it a single cherry-pick.
For commit only on master, it's still a good practice when the commit history conveys no information other than the developer fixing issues with the pull request, and making history/git blame needlessly hard to parse when doing maintenance. But it's not a strict requirement.

**Comment by aaime** [comment]
See also this:
http://softwareengineering.stackexchange.com/questions/263164/why-squash-git-commits-for-pull-requests
http://ndlib.github.io/practices/one-commit-per-pull-request/
http://eli.thegreenplace.net/2014/02/19/squashing-github-pull-requests-into-a-single-commit
http://codeinthehole.com/writing/pull-requests-and-other-good-practices-for-teams-using-github/

I see some developers trying to make a tidy history with unit changes and claiming it eases review, but at least personally, I find that a waste of time, the review should bother with the final result, not with the particular set of steps to get there (as there is normally no single "holy" path to do that).

**Comment by wumpz** [comment]
I should have said: GeoTools "strict" one commit policy. As you can see, the commits are distinct and one does not modify the others stuff. 

If you argue with time to do stuff, you should also consider the time someone needs to build a policy correct GeoTools pull request. As you may have noticed, this time I had to struggle. As for the review at GitHub one could use the "files changed" tab at the top to get the final changes that would be introduced. So no problem for a multicommit. 

Regarding linefeeds and tab/spaces stuff I am completly on your side. This time I had simply a configuration overlapping of two projects. So my fault.

Nevertheless I know the discussions for one / multi commit.

File: modules/library/main/src/main/java/org/geotools/styling/visitor/RescaleStyleVisitor.java [hunk 1]
```diff
@@ -133,6 +133,34 @@ protected Expression rescale( Expression expr ) {
...
+        List<Expression> rescaled = new ArrayList<>(expressions.size());
+        for (Expression expression : expressions) {
+            Expression rescale = ff.function("listMultiply", rescaleToExpression, expression);
+            if (expression instanceof Literal && evaluate) {
+                rescaled.add(ff.literal(rescale.evaluate(null)));
+            } else {
+                rescaled.add(rescale);
+            }
+        }
+        return rescaled;
```
Comment (wumpz) [inline modules/library/main/src/main/java/org/geotools/styling/visitor/RescaleStyleVisitor.java]: My intention was to precalculate all possible stuff. Otherwise one has to calculate it for every feature that hits this rule.

File: modules/library/main/src/test/java/org/geotools/styling/visitor/UomRescaleStyleVisitorTest.java [hunk 2]
```diff
@@ -293,6 +294,38 @@ protected void visitTextSymbolizerTest(double scaleMetersToPixel, Unit<Length> u
...
+			lineSymb.getStroke().setDashArray(Arrays.asList(filterFactory.literal("5.0 10.0")));
+
+            visitor = new UomRescaleStyleVisitor(scaleMetersToPixel);
+
+            lineSymb.accept(visitor);
+            LineSymbolizer rescaledLineSymb = (LineSymbolizer) visitor.getCopy();
+            String rescaledDynamicDashArray = (String)((Expression)rescaledLineSymb.getStroke().dashArray().get(0)).evaluate(null);
+            
+            assertEquals(expectedDashArray, rescaledDynamicDashArray);
+            assertNotSame(rescaledLineSymb, lineSymb);
```
Comment (wumpz) [inline modules/library/main/src/test/java/org/geotools/styling/visitor/UomRescaleStyleVisitorTest.java]: You are not going to force me to go another round, do you???

Comment (aaime) [inline modules/library/main/src/test/java/org/geotools/styling/visitor/UomRescaleStyleVisitorTest.java]: Yes, going round and round it's very frustrating. So is setting up [documented contribution rules](https://github.com/geotools/geotools/blob/master/CONTRIBUTING.md) and seeing them ignored. 

"Please make sure you're following the coding conventions, and otherwise avoid any reformats to the existing code, as they make it harder to review your changes.
[Coding conventions](http://docs.geotools.org/latest/developer/conventions/code/style.html) say:
"developers should use spaces for indentations, not tabulations"

Comment (wumpz) [inline modules/library/main/src/test/java/org/geotools/styling/visitor/UomRescaleStyleVisitorTest.java]: It is not frustrating for me to correct this simple misformattings but to provide this one commit thing, while it would be so easy to make another "correction" commit, while you could see the overall changes of all commits and could do a squash commit on github.

**Comment by wumpz** [comment]
@nmco: Do I see it right, that the only tabs regarding my commits are within dasharray3_dynamic.sld?

**Comment by nmco** [comment]
The file UomRescaleStyleVisitorTest.java contains tabs too.

**Comment by wumpz** [comment]
@nmco: Ah. Thx. But e.g. SLDStyleFactory.java  contains tabs within old code blocks. I am not supposed to correct those, or am I?

**Comment by nmco** [comment]
No you don't, you should only fix the ones introduced by the new code (i.e. this PR). Thanks.

**Comment by nmco** [comment]
Summarizing, the new contributed code should follow code policies (for example no tabs) and formatting issue should fixed in the old code.

**Comment by wumpz** [comment]
@nmco: The changes are there. One more question regarding a comment about squashing from above:

> Regarding the commit squashing, it can be done by the merger but the commit will look like it's done
> by the merger not by you. You will lose attribution of this work.

What do you mean by saying "will lose attribution of this work"? As far as I know, github will preserve the committer and will build a summarized message from all processed commits. Or did I missunderstand you?

**Comment by wumpz** [comment]
@nmco: Do you have a tool configured within the geotools project, like checkstyle, to check project formatting issues?

**Comment by nmco** [comment]
In the documentation there is a section that talks about that (Use of Formatting Tools):
http://docs.geotools.org/latest/developer/conventions/code/style.html

**Comment by nmco** [comment]
Someone reported that the commit looks like is done by the merger not he original author.

**Comment by wumpz** [comment]
@nmco: I have seen this, and thought it was all right set up, but it unfortunatly was not. 

My question was regarding a tool to check not to format by itself.  It would be cool to have a e.g. checkstyle maven plugin configured, to check all that. Sure it would be nessecary to make it in a different profile, as we could see, that the "old" code doesnt follow the convention partially. But a future configuration could force this check step on every build and change.

**Comment by nmco** [comment]
Right, there was discussion about this in the mailing list. At the moment no there is not something like that.

**Comment by bencaradocdavies** [comment]
I like to squash locally so I can use ```git commit --author``` to preserve the original authorship.

**Comment by wumpz** [comment]
But does github merge squash not the same?

**Comment by bencaradocdavies** [comment]
[About merge methods on GitHub](https://help.github.com/articles/about-merge-methods-on-github/): "Before enabling squashing commits, consider these disadvantages: - You lose information about when specific changes were originally made and who authored the squashed commits. [...]"

**Comment by bencaradocdavies** [comment]
And if a merge squashes commits from several authors, what should GitHub do?

If I recall correctly, squashed commits are just a single commit from the merger. With the title from the PR? No sign of the original author. Moral of the story: preserve your moral right of attribution by making tidy pull requests. Make your pull request easy to read and easy to (non-squash) merge, with a single commit per topic and a good message with a Jira issue in each commit.

Pull requests refer to the source branch, so it is easy to update them with tidy commits by for-pushing better commits to the source branch.

**Comment by wumpz** [comment]
So is there anything else to do? Or missed I something?
